src/stdlib/replace_with.rs

```diff
@@ -0,0 +1,214 @@
+{
+    let haystack = value.try_bytes_utf8_lossy()?;
+    let count = match count.try_integer()? {
+        i if i > 0 => i as usize,
+        i if i < 0 => 0,
+        // this is when i == 0
+        _ => return Ok(value),
+    };
+    match pattern {
+        Value::Regex(regex) => {
+            let captures = regex.captures_iter(&haystack);
+            make_replacement(captures, &haystack, count, ctx, runner)
+        }
+        value => Err(ValueError::Expected {
+            got: value.kind(),
+            expected: Kind::regex(),
+        }
+        .into()),
+    }
+}
+
+fn make_replacement<T>(
+    caps: regex::CaptureMatches,
+    haystack: &str,
+    count: usize,
+    ctx: &mut Context,
+    runner: closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Result<Value, ExpressionError>,
+{
+    // possible optimization: peek at first capture, if none return the original value.
+    let mut replaced = String::with_capacity(haystack.len());
+    let limit = if count == 0 { usize::MAX } else { count - 1 };
+    let mut last_match = 0;
+    // we loop over the matches ourselves instead of calling Regex::replacen, so that we can
+    // handle errors. This is however based on the implementation of Regex::replacen
+    for (i, cap) in caps.enumerate() {
+        // Safe to unrap because the 0th index always includes the full match.
+        let m = cap.get(0).unwrap(); // full match
+
+        let mut value = captures_to_value(&cap);
+        runner.map_value(ctx, &mut value)?;
+        let replacement = value.try_bytes_utf8_lossy()?;
+
+        replaced.push_str(&haystack[last_match..m.start()]);
+        replaced.push_str(&replacement);
+        last_match = m.end();
+        if i >= limit {
+            break;
+        }
+    }
+    // add the final component
+    replaced.push_str(&haystack[last_match..]);
+    Ok(replaced.into())
+}
+
+fn captures_to_value(captures: &regex::Captures) -> Value {
+    // return an array of the capture groups
+    captures
+        .iter()
+        .map(|m| match m {
+            Some(m) => m.as_str().into(),
+            None => Value::Null,
+        })
+        .collect()
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct ReplaceWith;
+
+impl Function for ReplaceWith {
+    fn identifier(&self) -> &'static str {
+        "replace_with"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "pattern",
+                kind: kind::REGEX,
+                required: true,
+            },
+            Parameter {
+                keyword: "count",
+                kind: kind::INTEGER,
+                required: false,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
```

Comment:
We will also need a PR in the Vector repo because that's where we generate the public docs (hopefully that can be simplified in the future). See [this for example](https://github.com/vectordotdev/vector/blob/master/website/cue/reference/remap/functions/replace.cue).

Reply:
I created https://github.com/vectordotdev/vector/pull/19638

Reply:
Thanks! I will review/merge that after the next VRL release.